### ISSUE TITLE
Drush site-set fails for aliases in site specific locations (e.g. sites/all/drush)

### ISSUE THREAD

**Issue by alex-mccabe** [drush-ops/drush#1532]
#1241 may or may not be related, I'm not sure. In that issue aliases stopped working entirely, in here they only stop working with site-set, but in both cases it breaks at 7.0.0-alpha4 - alpha3 works fine.

Running a command with an explicit alias works:

`$ drush @mysite.local en -y devel`
`devel is already enabled.                                                                                                                                                                                                  [ok]`
`There were no extensions that could be enabled.                                                                                                                                                                            [ok]`

Running site-set (or the alias, use) and then the command without the alias fails:

`$ drush use @mysite.local`
`$ drush en -y devel`
`Command pm-enable needs a higher bootstrap level to run - you will need to invoke drush from a more functional Drupal environment to run this command.                                                              [error]`
`The drush command 'en devel' could not be executed.                                                                                                                                                                 [error]`
`Could not find a Drupal settings.php file at sites/default/settings.php.                                                                                                                                            [error]`

The alias is for a local development site that has its own settings.php in its own directory, `mysite.local/settings.php`. I'm on Ubuntu 14.04 LTS, Drupal 7.38, running PHP 5.5.9-1ubuntu4.11.

**Comment by weitzman** [comment]
Please report bugs using master branch or at least a released version of 7.x

**Comment by alex-mccabe** [comment]
@weitzman I'm not sure I follow. This happens in 7.0.0-alpha4 and higher, *including* 7.0.0, and dev-master.

Or is there something that you're asking me to do here in GitHub to attach it to one of those branches?

**Comment by weitzman** [comment]
Thanks for clarifying.

**Comment by weitzman** [comment]
Can you please provide the output of `drush sa @mysite.local`.

**Comment by joachim-n** [comment]
I am seeing the same problem. Aliases work with commands such as 'status' and 'en' and 'fl', but while site-set reports success, it has no lasting effect.

Here's the requested output for my alias:

```
drush sa @goh
$aliases["goh"] = array (
  'uri' => 'http://localhost:8888/mysite/subsite/',
  'path-aliases' =>
  array (
  ),
);
```

**Comment by ohthehugemanatee** [comment]
I'm having the same problem. Aliases file is in ```sites/all/drush/foo.aliases.drushrc.php``` :

```php
<?php

// Figure out where the vagrant private key is stored (it moved in a recent version)
$local_root = drush_get_context('DRUSH_SELECTED_DRUPAL_ROOT');
if (is_file($local_root . '/.vagrant.d/insecure_private_key')) {
  $insecure_private_key = $local_root . '/.vagrant.d/insecure_private_key';
}
else {
  $insecure_private_key = '~/.vagrant.d/insecure_private_key';
}

$aliases['local'] = array(
  'parent' => '@parent',
  'uri' => 'http://10.11.12.14',
  'root' => '/vagrant/public',
  'remote-host' => '10.11.12.14',
  'remote-user' => 'vagrant',
  'ssh-options' => "-i " . $insecure_private_key . " -l vagrant",
  'db-url' => 'mysql://web:web@10.11.12.14:3306/web',
  'databases' => array (
    'default' => array (
      'default' => array (
        'database' => 'web',
        'username' => 'web',
        'password' => 'web',
        'host' => '10.11.12.14',
        'port' => '',
        'driver' => 'mysql',
        'prefix' => '',
      ),
    ),
  ),
);

$aliases['dev'] = array(
  'uri' => 'http://dev.foo.byf1.io',
  'root' => '/var/www/vhosts/foo.dev/public',
  'remote-host' => 'dev.foo.byf1.io',
  'remote-user' => 'foo',
  'command-specific' => array(
    'rsync' => array ( 
      'mode' => 'rzv'
    ),
  ),  
);

$aliases['stage'] = array(
  'uri' => 'http://stage.foo.byf1.io',
  'root' => '/var/www/vhosts/foo.stage/public',
  'remote-host' => 'stage.foo.byf1.io',
  'remote-user' => 'foo',
  'command-specific' => array(
    'rsync' => array (
      'mode' => 'rzv'
    ),
  ),
);
```

The aliases all work fine if I specify them on the command line, ie ``` drush @foo.dev sqlc```. 

```drush use @foo.dev``` reports success.

But

``` bash
~/Sites/foo/public $ drush sa @foo.dev
$aliases["foo.dev"] = array (
  'uri' => 'http://dev.foo.byf1.io',
  'root' => '/var/www/vhosts/foo.dev/public',
  'remote-host' => 'dev.foo.byf1.io',
  'remote-user' => 'foo',
  'command-specific' => 
  array (
    'rsync' => 
    array (
      'mode' => 'rzv',
    ),
  ),
);
~/Sites/foo/public $ drush use @foo.dev
Site set to @foo.dev
~/Sites/foo/public $ drush sqlc
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/Applications/MAMP/tmp/mysql/mysql.sock' (2)
~/Sites/foo/public $
```

I tried running with ```--debug```, but it just shows where it scans for alias files.

**Comment by ohthehugemanatee** [comment]
Did some more testing today, and got more information:
-  Everything works if the aliases file is in ~/.drush ... So it's some kind of problem with the file location sites/all/drush
- I tried adding an echo statement to the top of my aliases file to see when it's getting parsed. It's running 2x both when ```drush use``` works, and when it doesn't.
- I tried some echo debugging, and found out where the failure happens: In [includes/sitealias.inc:492](https://github.com/drush-ops/drush/blob/master/includes/sitealias.inc#L492), we try to match the name of the "site-set" provided alias to a known alias with ```drush_get_context('site-aliases')```. This fails to return any of the aliases from ```sites/all/drush```. 

**TL;DR** when I run ```drush status``` the site-aliases context does not get the contents of ```sites/all/drush/foo.aliases.drushrc.php```.  When I run ```drush sa``` it is properly populated. 

I tried poking through the context loading process, but it's quite complicated to just jump in the middle. I'm happy to take this further if someone can give me a direction. Where do most commands build their context, and how is that different from where ```drush sa``` builds its context?

**Comment by weitzman** [comment]
Thanks for the helpful debugging.

Very early during preflight, we look at the file containing the alias that you specified in site-set. The code that reads this file is [drush_sitealias_check_arg_and_site_set()](https://github.com/drush-ops/drush/blob/8.0.1/includes/sitealias.inc#L34). As you can see from that comment, we silently ignore any specified alias which is not recognized. Aliases specified in sites/all/drush have not been recognized yet. Those get recognized a few lines later, in the call to drush_preflight_site(). So if we could defer or duplicate the call to drush_sitealias_check_arg_and_site_set() a bit later, we will have fixed this buglet. Ping to @greg-1-anderson for feasibility of that. 

if you want to use site-set today, you would have to specify --alias-path=/path/to/sites/all/drush. I acknowledge that this not very convenient. I'm sure that the silent failure is a bit frustrating as well.

**Comment by greg-1-anderson** [comment]
It does seem to me that we should complain if we cannot find an alias that site-set tried to select.

Site-local aliases are a horrible, horrible feature. They were intended to allow a given site to specify where other copies of itself reside, e.g. the remote live and staging servers. However, that particular use-case is almost never needed. Typically, folks get excited that they can share their site-local aliases in the git repository that holds the site. This is a powerful motivator for using site-local aliases, but unfortunately, folks usually get hung up by the fact that Drush needs to find your site in order to find your site-local aliases, so using a site-local alias to find the site that contains it is a non-starter.

Some folks try to use site-local aliases in conjunction with cwd-based site selection. This sort of works, but runs into all sorts of edge cases (e.g. above). This use case is also unnecessary; if you have already selected your Drupal root from the cwd, then you can specify the multisite you want to use via `drush --uri=site status` or `drush #site status`.  Any site-specific options you want set can then be specified in a drushrc.php file located in the same directory that contains settings.php, which is a much more stable choice than putting the same options in a site-local alias.

Referencing DRUSH_SELECTED_DRUPAL_ROOT from within an alias file is similarly dubious. Could this path be found any other way, e.g. relative to HOME?

I don't intend to blame the people who use site-local aliases here; it happens time and time again, so it is clear that it's natural for folks to presume that this use-case should be supported.  I'd love to figure out what we could change or make clearer to keep people from falling into this trap, though. I'd prefer to try to fix this in a larger way (e.g. by removing site-local aliases) rather than fiddle with bootstrap order, but I will look into the suggestion above and see if it might be viable to make that adjustment without breaking something else.

**Comment by greg-1-anderson** [comment]
I pushed a couple of potential solutions, but did not test them.

It would be theoretically possible to detect at `site-set` time that the found alias was site-local. In that instance, we could set not the alias name, but `path/to/root/#uri`, which would always be found. We'd want to try to find and load the alias later, in case it set options & c.

I don't feel warmly enough towards site-local aliases to try to fix it with this much code, though.

**Comment by weitzman** [comment]
If folks are able to test #1902, that would be helpful.

**Comment by ohthehugemanatee** [comment]
@weitzman thanks for the explanation, and @greg-1-anderson your comment explains a lot about the Right Way To Do It mindset. Thank you!

I'm an experienced drupalist and always considered myself a drush power-user, and the only way *I* knew of to bundle connections to dev/staging/live into the repo was in a site-local alias. It never occurred to me to try doing it any other way. @greg-1-anderson's comment gives me hope that maybe we can deprecate sites/all/drush as an alias files location and address this whole use case with documentation.

My requirements (which I think are common) are:
- allow drush to operate on remote installations with configuration 100% contained in the repo.
- do not require maintaining aliases as a separate step from maintaining cloned repos. (ie by copying alias files around)

I've always done this with site-local aliases and cwd-based site selection, but I'm open to options. Some alternatives off the top of my head:
- a drushrc.php include line that looks for alias files inside ~/Sites/* , so aliases ANYWHERE in my projects folder are included
- cwd-based site detection, with a "multisite" site for each remote (sites/dev/ , sites/stage/ , etc) that contains DB connection info in settings.php, and SSH connection info in drushrc.php (that's possible, right?)

If I understand your explained solution, it's the latter option. But it seems odd to use sites/* to do the job of remote aliases. Are there other, better ways to do this?

**Comment by joachim-n** [comment]
By site-local aliases, do you mean ones defined in DRUPAL/sites/all/drush/aliases.drushrc.php and also ones in DRUPAL/drush/aliases.drushrc.php?

Defining aliases per-codebase is extremely handy, as without that, you'd have all your aliases available in every site. 'drush sa' would be much harder to use when trying to remember which alias to use. (I typically have about 5-6 aliases for local multisites, then test/dev/live; on Pantheon sites there are even more.)

**Comment by greg-1-anderson** [comment]
Yes, site-local aliases are defined inside a site's document root, such as sites/all/drush/aliases.drushrc.php.

The intended use-case for site-local aliases is to store remote aliases associated with a site. This use-case is still valid, and has no good alternative mechanism, so we should not go so far as to remove this feature.

Adding an include line to your global drushrc.php that adds ~/Sites/*/sites/all/drush as an alias search path works today.  If you store all of your sites in ~/Sites, this would effectively convert all of your site-local aliases into global aliases.  This would allow you to manage your aliases in your sites' repositories, and give you all of the advantages of ordinary global aliases; however, it would not address @joachim-n's desire to avoid unwieldy output from `drush sa`.

If you only have one settings.php in each site (in default), then you can just use `drush status` & c. (no alias) with cwd site selection for the local site.  This is supported.

If you do use multisites, try using `drush #foo status` instead of an alias, and see if that works for you. As I mentioned above, your local options for the foo site could then go in `sites/foo/drushrc.php`. This should work, but I don't know if a lot of people are using it, or if we have tests for it. I'd be happy to fix any problems encountered with this feature if this is a workable solution for folks.

If you want to use `use @mysite.remote` with cwd-based site selection, and remote site aliases, try the fix in #1902 and report back as to whether it helps you or not.

If we want to support `use @mysite.local` or `use #foo` or `use @self`, followed by `cd /some/other/location`, then we would need a more complex solution such as the [one I mentioned above](https://github.com/drush-ops/drush/issues/1532#issuecomment-169788070).

**Comment by ohthehugemanatee** [comment]
tested the fix from #1902 and it works for me. Thanks for the clear thinking, explanations, and code!

**Comment by weitzman** [comment]
Fixed by #1902

**Comment by weitzman** [comment]
If people want to store aliases in their codebase (I like this approach as well), one good solution is to copy and customize [drush.wrapper](https://github.com/drush-ops/drush/blob/master/examples/drush.wrapper). The relevant customization is to add --alias-path. Note that this approach does not work when using drush.phar, but these folks should put Drush itself into their repo.

Given the above, perhaps we can remove support for site-local aliases?

**Comment by greg-1-anderson** [comment]
Unless I am mistaken, the `drush.finder` should work just fine if the global and/or site-local Drush are a phar.  The only thing you lose in a phar are the features of the `drush` bash script, which includes selection of the php executable, php.ini management, et. al.

Using the `drush.wrapper`, the aliases stored in your site are only visible to Drush if said site is selected by the drush.finder, e.g. via cwd site selection. The drush wrapper can be committed to your site's repository, so alias sharing in teams is supported. However, I don't think that we can remove site-local alias support, because the `drush.wrapper` technique also depends on the fix in #1902 to support `use` with a site-local alias.  Specifying the location of your alias files via `drush.wrapper` is equivalent to using the default location vis-a-vis behavior of site-local aliases.

Maybe I need to give up, and learn to love the site-local alias. My workflow is to live with the cumbersome `site-aliases` list that comes with pure global alias management strategy. This workflow is not feasible if you have hundreds of sites, and is inconvenient if you are a member of multiple teams.

For the record, the thing I hate the most about site-local aliases is the need to re-run the same bootstrap phases multiple times. In the current code, this wart is isolated to [one block at the end of drush_preflight()](https://github.com/drush-ops/drush/blob/8bcabc9ab324bbeebc6e9e65b45afc92ccf48f05/includes/preflight.inc#L275), and another location where we force a call to drush_invoke_process if it looks like we're about to get into trouble (i.e. if someone uses a site-local alias that points to a different Drupal site on the same machine). If we run drush_preflight_root() and / or drush_preflight_site() twice (as we do in the code referenced above), we run the risk of blowing up when we load the second site's autoload file. This hasn't been a practical problem since we separated the bootstrap from the preflight (in Drush 5, I think).

So, perhaps we're okay, with just a bit of ugly code that we need to live with.

**Comment by greg-1-anderson** [comment]
I cherry-picked #1902 onto the 7.x branch.

**Comment by weitzman** [comment]
Unless I am mistaken, the drush.finder should work just fine if the global and/or site-local Drush are a phar.

Right. I stand corrected.

    Specifying the location of your alias files via drush.wrapper is equivalent to using the default location vis-a-vis behavior of site-local aliases.

You are correct again. --alias-path happens to be read at end of drush_preflight_site() when we first create a @self site alias. 

I too dislike the re-running of bootstrap steps.

Nothing more is actionable here AFAICT so closing.
